---
title: "Combined A673 tumor scRNAseq with HSV reference"
author: "Emily Ringwalt"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 5
    number_sections: false
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  echo = TRUE,
  cache = TRUE,
  collapse = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  cache.lazy = FALSE)
```

```{r lib, cache = FALSE}
library(Seurat)
library(tidyverse)
library(ggsci)
library(rrrSingleCellUtils)
library(patchwork)
library(data.table)
library(enrichplot)
source("scSeurat-addhsv.R")

# Create output directories as needed
my_dirs <- c("Data",
             "Plots",
             "Plots/A673_comb")

lapply(my_dirs[file.exists(my_dirs) == FALSE], function(x) dir.create(x))
```

# Combine A673 Tumor Samples

## Normalize and Process

```{r process, fig.height=3, fig.width=3}
# Set seed
set.seed(888)

# Load the two A673 datasets
## From 'HSVref_A673.Rmd' and 'HSVref_A673_human-immune-samples.Rmd
load("Data/tumor-only-A673_hsv_rep1_even-labeled.RData")
load("Data/tumor-immune-A673_hsv_rep2_even-labeled.RData")

# Merge and process using rrrSingleCellUtils package
A673_hsv_comb <-
  merge(A673_hsv_rep1_even,
        A673_hsv_rep2_even,
        add.cell.ids = c("A673_tumoronly",
                         "A673_tumorall"),
        project = "Trabectedin plus HSV1716 in A673 Combined Datasets") %>%
  process_seurat(resolution = 0.1,
                 run_umap_dims = 1:18)

# use ElbowPlot to determine PC to include
ElbowPlot(A673_hsv_comb,
          ndims = 30)

# Remove individual objects once merged
rm(A673_hsv_rep1_even)
rm(A673_hsv_rep2_even)
```

### Clustering Resolution {.tabset}

#### Res0.1

```{r umap, fig.height=5, fig.width=10, dependson=c("load_final","process")}
# Use resolution of 0.1 from above
DimPlot(A673_hsv_comb,
        reduction = "umap",
        split.by = "treatment",
        pt.size = 1,
        label = T) +
  coord_fixed()
```

#### Res0.2

```{r umap2, fig.height=5, fig.width=10, dependson="umap"}
A673_hsv_comb <- FindClusters(A673_hsv_comb,
                           resolution = 0.2)

DimPlot(A673_hsv_comb,
        reduction = "umap",
        split.by = "treatment",
        pt.size = 1,
        label = T) +
  coord_fixed()
```

#### Res0.3

```{r umap3, fig.height=5, fig.width=10, dependson="umap"}
A673_hsv_comb <- FindClusters(A673_hsv_comb,
                           resolution = 0.3)

DimPlot(A673_hsv_comb,
        reduction = "umap",
        split.by = "treatment",
        pt.size = 1,
        label = T) +
  coord_fixed()
```

### DGEA by Cluster

Differentially Expressed Genes by Cluster (res0.3):

```{r deg, dependson=c("load_final","process","umap")}
A673_hsv_comb_markers <- FindAllMarkers(A673_hsv_comb,
                                        min.pct = 0.25,
                                        logfc.threshold = 0.25,
                                        only.pos = T)
```

```{r degt, dependson="deg"}
# Isolate top 10 in each cluster
A673_hsv_comb_markers %>%
  group_by(cluster) %>%
  top_n(n = 10, wt = avg_log2FC) -> t

# Get dataframe into proper format
t <- as.data.frame(t)
t$cluster <- as.numeric(as.character(t$cluster)) # mutate
t <- t[order(t$cluster),] #arrange
t <- t[,c(7, 1:6)] # relocate
t <- t[,c(7, 1:6)] # relocate

# Create interactive datatable showing top 10 DEG in each cluster 
DT::datatable(t, 
              rownames = FALSE,
              extensions = c('FixedColumns',
                             'Buttons'),
              options = list(
                pageLength = 10,
                scrollX = TRUE,
                scrollCollapse = TRUE,
                dom = 'Bfrtip',
                buttons = c('copy',
                            'csv',
                            'excel')
              ))
```

Check that there is the same number of cells in each treatment group.

```{r downsample, fig.height=5, fig.width=10, dependson=c("load_final","process","umap")}
# Check that cell numbers in each treatment group are equal
## Extract meta data
library(data.table)
A673_hsv_comb.d <- A673_hsv_comb@meta.data %>%
  as.data.table
## Resulting table has one "row" per cell - count
A673_hsv_comb.d[, .N, by = "treatment"]

# Rename to indicate that treatment group numbers are equal ("even")
A673_hsv_comb_even <- A673_hsv_comb

# Remove extra object
rm(A673_hsv_comb)
```

# oHSV Analysis

```{r matthelp_hsvgenes, dependson = c("load_final","process","umap","downsample")}
# Find head/prefix ^HSV1 in all genes (rownames) of tumor object
# Output is a vector of genes beginning with HSV1 that were found in the data:
hsv <- str_subset(row.names(A673_hsv_comb_even),
                  pattern = "^HSV1")
```

**All oHSV Genes Present:**

```{r hsv_genes, dependson = "matthelp_hsvgenes"}
# Create interactive datatable showing hsv genes
hsv_dt <- as.data.table(hsv)
DT::datatable(hsv_dt, 
              rownames = FALSE,
              extensions = c('FixedColumns',
                             'Buttons'),
              options = list(
                pageLength = 10,
                scrollX = TRUE,
                scrollCollapse = TRUE,
                dom = 'Bfrtip',
                buttons = c('copy',
                            'csv',
                            'excel')
              ))
```

A closer look at HSV transcript presence in the data:

```{r matthelp_assaydata, dependson = "matthelp_hsvgenes"}
# Make a compressed matrix of HSV1 genes as rownames and the corresponding 
#single cell data for each gene (getassaydata is scaled data)
assay.data <- GetAssayData(A673_hsv_comb_even[hsv])

# Rownames of compressed matrix (assay.data) should be the genes from the subset above
#row.names(assay.data)

# Make a numeric vector of the column sums 
assay_col <- colSums(assay.data)

assay_col_unscaled <- GetAssayData(A673_hsv_comb_even[hsv], ) %>%
  colSums()
(assay_col > 0) %>%
  summary()
```

Distribution of HSV1 transcripts:

```{r tibble_hist, dependson = c("matthelp_hsvgenes","matthelp_assaydata")}
#Use tibble and ggplot to look at histogram distribution of HSV genes 
tibble(cell_sum = assay_col) %>%
  ggplot(aes(cell_sum)) +
  geom_histogram(bins = 200)
```

Log normalized distribution of HSV1 transcripts:

```{r tibble_hist_log, dependson = c("matthelp_hsvgenes","matthelp_assaydata")}
#log scale helps to separate out low expression values and zero. We add 0.000001 to make sure we don't have log(0) (equals infinity so algorithm will remove these values) but can still view all the data points. 
tibble(cell_sum = assay_col) %>%
  ggplot(aes(cell_sum + 0.000001)) +
  geom_histogram(bins = 200) +
  scale_x_log10()
```

Given that this analysis captures viral transcript presence/abundance well, 
it is added to the tumor data as metadata. 

```{r new_hsv_meta, dependson = c("matthelp_hsvgenes","matthelp_assaydata")}
A673_hsv_comb_even <- AddMetaData(A673_hsv_comb_even,
                               assay_col,
                               col.name = "oHSV_abundance")

A673_hsv_comb_even <- AddMetaData(A673_hsv_comb_even,
                          assay_col > 0,
                          col.name = "oHSV_presence")
```

### DimPlots: oHSV presence
*Figure 2C*

```{r hsvdimplot, fig.height=5, fig.width=10, dependson = c("matthelp_hsvgenes","matthelp_assaydata","new_hsv_meta")}

#### Figure 2C #### 
Idents(A673_hsv_comb_even) <- A673_hsv_comb_even$oHSV_presence

# Change identities from TRUE/FALSE to infected/uninfected 
A673_hsv_comb_even <- RenameIdents(A673_hsv_comb_even, 
                                c("TRUE" = "oHSV+", 
                                  "FALSE" = "oHSV-"))

DimPlot(A673_hsv_comb_even,
        pt.size = 0.5,
        order = T) +
  ggtitle("oHSV Transcript Presence") +
  coord_fixed()

ohsvpresence <- DimPlot(A673_hsv_comb_even,
                        pt.size = 0.5,
                        split.by = "treatment",
                        order = TRUE) +
  ggtitle("oHSV Transcript Presence") +
  coord_fixed()
ohsvpresence

pdf("Plots/A673_comb/Figure2C_ohsv_presence_dimplot.pdf",
    height = 5,
    width = 10)
ohsvpresence
dev.off()
```

### FeaturePlots: oHSV abundance

```{r featurehsv, dependson = c("matthelp_hsvgenes","matthelp_assaydata","new_hsv_meta")}
FeaturePlot(A673_hsv_comb_even,
            features = "oHSV_abundance",
            order = TRUE) +
  theme(aspect.ratio = 1)
```

**Split by treatment group**

```{r feature_trt_hsv, fig.height = 3, fig.width = 12, dependson = c("matthelp_hsvgenes","matthelp_assaydata","new_hsv_meta")}
FeaturePlot(A673_hsv_comb_even,
            split.by="treatment",
            features = "oHSV_abundance",
            order = TRUE) +
  theme(aspect.ratio = 1)
```

**Max Cutoff at 100**

```{r feature_trt_hsv_100, fig.height = 3, fig.width = 12, dependson = c("matthelp_hsvgenes","matthelp_assaydata","new_hsv_meta")}
# Max cutoff at 100
FeaturePlot(A673_hsv_comb_even,
            split.by="treatment",
            features = "oHSV_abundance",
            order = TRUE,
            max.cutoff = 100) +
  theme(aspect.ratio = 1)
```

**Max Cutoff at 50**

```{r feature_trt_hsv_50, fig.height = 3, fig.width = 12, dependson = c("matthelp_hsvgenes","matthelp_assaydata","new_hsv_meta")}
# Max cutoff at 50
FeaturePlot(A673_hsv_comb_even,
            split.by="treatment",
            features = "oHSV_abundance",
            order = TRUE,
            max.cutoff = 50) +
  theme(aspect.ratio = 1)
```

**Max Cutoff at 20**

```{r feature_trt_hsv_20, fig.height = 3, fig.width = 12, dependson = c("matthelp_hsvgenes","matthelp_assaydata","new_hsv_meta")}
# Max cutoff at 50
FeaturePlot(A673_hsv_comb_even,
            split.by="treatment",
            features = "oHSV_abundance",
            order = TRUE,
            max.cutoff = 20) +
  theme(aspect.ratio = 1)
```

### Violin Plots: oHSV abundance

oHSV abundance is shown below across treatment groups, as well as by cluster and treatment groups.

```{r hsv_ab, dependson = c("matthelp_hsvgenes","matthelp_assaydata","new_hsv_meta")}
Idents(A673_hsv_comb_even) <- A673_hsv_comb_even$treatment

# Violin plots of HSV abundance
VlnPlot(A673_hsv_comb_even,
        features = "oHSV_abundance",
        split.by = "treatment",
        cols = c("grey", "darkblue", "orangered", "purple"))

# Simplify group names for later editing in Illustrator
obj4vln <- RenameIdents(A673_hsv_comb_even,
                        c("1 - Control" = "1",
                          "2 - oHSV" = "2",
                          "3 - Trabectedin" = "3",
                          "4 - oHSV+Trabectedin" = "4"))

#### Figure 2D ####
# Violin plot with new naming and on log scale
log_ab <- VlnPlot(obj4vln,
                  features = "oHSV_abundance",
                  split.by = "treatment",
                  log = TRUE,
                  cols = c("grey", "darkblue", "orangered", "purple")) +
  NoLegend()
# Plot in rmd html output
log_ab

# Save as pdf
pdf("Plots/A673_comb/Figure2D_vln_log_hsvabundance.pdf",
    height = 5, width = 5)
print(log_ab)
dev.off()
```

### HSV Time Point Module Scores

Look at immediate early, early, and late gene expression and add to the data as a module score. Below the baseline plot, cutoffs are used to better visualize the spread of oHSV transcripts for each stage of gene expression. Ensuring that the expression patterns are not exactly the same supports that presence of oHSV transcripts is not due to random spread during 10x sample preparation.

Immediate early: "HSV1-RL2", "HSV1-RS1", "HSV1-UL54", "HSV1-US1", "HSV1-US12"

Early: "HSV1-UL23", "HSV1-UL29", "HSV1-UL50", "HSV1-UL2"

Late: "HSV1-UL48", "HSV1-UL19", "HSV1-US6" (note: not present in our data), "HSV1-UL27", "HSV1-UL53", "HSV1-UL44", "HSV1-UL41"

```{r hsv_timepoints, fig.height = 4, fig.width = 16, dependson = c("matthelp_hsvgenes","matthelp_assaydata","new_hsv_meta")}
# Create lists of time points for HSV gene expression 
imm_early <- list(c("HSV1-RL2",
                    "HSV1-RS1",
                    "HSV1-UL54",
                    "HSV1-US1",
                    "HSV1-US12"))

early <- list(c("HSV1-UL23",
                "HSV1-UL29",
                "HSV1-UL50",
                "HSV1-UL2"))

late <- list(c("HSV1-UL48",
               "HSV1-UL19",
               "HSV1-US6",
               "HSV1-UL27",
               "HSV1-UL53",
               "HSV1-UL44",
               "HSV1-UL41"))

# Add HSV timepoint module scores 
timepoints <- c("Immediate_Early_Genes",
                "Early_Genes",
                "Late_Genes")

hsv_timepoints <- c(imm_early,
                    early,
                    late)
names(hsv_timepoints) <- timepoints

hsv_timepoints <-
  list("Immediate_Early_Genes" = c("HSV1-RL2",
                                   "HSV1-RS1",
                                   "HSV1-UL54",
                                   "HSV1-US1",
                                   "HSV1-US12"),
       "Early_Genes"           = c("HSV1-UL23",
                                   "HSV1-UL29",
                                   "HSV1-UL50",
                                   "HSV1-UL2"),
       "Late_Genes"            = c("HSV1-UL48",
                                   "HSV1-UL19",
                                   "HSV1-US6",
                                   "HSV1-UL27",
                                   "HSV1-UL53",
                                   "HSV1-UL44",
                                   "HSV1-UL41"))

plots <- list()
for (hsv_time in names(hsv_timepoints)) {
  # Add timepoint data as module score
  A673_hsv_comb_even <- AddModuleScore(A673_hsv_comb_even,
                                       features = list(hsv_timepoints[[hsv_time]]),
                                       ctrl = 5,
                                       name = hsv_time)
  
  # Rename AddModuleScore naming syntax
  colnames(A673_hsv_comb_even@meta.data)[which(names(A673_hsv_comb_even@meta.data) ==
                                                 paste0(hsv_time, "1"))] <-
    hsv_time
  
  # Create FeaturePlots based on the timepoint module scores
  plots[[paste0(hsv_time, "_plot")]] <- 
    r_feature_plot(A673_hsv_comb_even,
                   features = hsv_time,
                   split.by = "treatment",
                   order = TRUE,
                   min.cutoff = 0) +
    theme(aspect.ratio = 1)
  
  print(plots[[paste0(hsv_time, "_plot")]])

# Create FeaturePlots based on the timepoint module scores
  ## Cutoff of 15
  plots[[paste0(hsv_time, "_plot_cutoff15")]] <- 
    r_feature_plot(A673_hsv_comb_even,
                   features = hsv_time,
                   split.by = "treatment",
                   order = TRUE,
                   max.cutoff = 15,
                   min.cutoff = 0) +
    theme(aspect.ratio = 1)
  print("Max Cutoff of 15")
  print(plots[[paste0(hsv_time, "_plot_cutoff15")]])
  
  ## Cutoff of 1
  plots[[paste0(hsv_time, "_plot_cutoff1")]] <- 
    r_feature_plot(A673_hsv_comb_even,
                   features = hsv_time,
                   split.by = "treatment",
                   order = TRUE,
                   max.cutoff = 1,
                   min.cutoff = 0) +
    theme(aspect.ratio = 1)
  print("Max Cutoff of 1")
  print(plots[[paste0(hsv_time, "_plot_cutoff1")]])
  
  pdf(paste0("Plots/A673_comb/rfeaturecutoff1_",
             hsv_time,
             ".pdf"),
      height = 5,
      width = 10)
  print(plots[[paste0(hsv_time, "rfeatureplot_plot_cutoff1")]])
  dev.off()
}

save(A673_hsv_comb_even,
     file  = "Data/A673_hsv_comb_even-labeled.RData")
```

*Figure 2E*

```{r hsv-dot, height = 10, width = 10, dependson = "hsv_timepoints"}
#### Figure 2E ####
# Set cell identities to treatment group
Idents(A673_hsv_comb_even) <- A673_hsv_comb_even$treatment

# Feature plot of timepoint gene expression for groups of interest
feature <- subset(A673_hsv_comb_even,
                  idents = c("2 - oHSV", "4 - oHSV+Trabectedin")) %>%
  r_feature_plot(features = c("Immediate_Early_Genes",
                              "Early_Genes",
                              "Late_Genes"),
                 split.by = "treatment",
                 order = TRUE,
                 max.cutoff = 1,
                 min.cutoff = 0) +
  theme(aspect.ratio = 1)
print(feature)

# Save as pdf
pdf("Plots/A673_comb/Figure2E_featureplot_time.pdf",
    height = 12.5,
    width = 8.25)
print(feature)
dev.off()

# Violin plot of timepoint gene expression for groups of interest
time_vln <- subset(A673_hsv_comb_even,
                   idents = c("2 - oHSV", "4 - oHSV+Trabectedin")) %>%
  VlnPlot(features = c("Immediate_Early_Genes",
                       "Early_Genes",
                       "Late_Genes"),
          group.by = "treatment",
          split.by = "treatment",
          same.y.lims = T,
          log = T,
          cols = c("darkblue", "purple")) +
  NoLegend()
print(time_vln)

# Save as pdf
pdf("Plots/A673_comb/Figure2F_vlnplot_time.pdf",
    height = 5,
    width = 9)
print(time_vln)
dev.off()
```

*Figure S4A*

```{r timeexp, dependson = "hsv_timepoints"}
#### Figure S4A ####
# Isolate the time-related HSV genes found in the A673 data
hsv_genes <- intersect(c(imm_early[[1]],
                         early[[1]],
                         late[[1]]),
                       rownames(A673_hsv_comb_even))

# Calculate fold change/p values for HSV1 genes in the A673 data
hsv_DEG <- FindMarkers(A673_hsv_comb_even,
                       ident.1 = "4 - oHSV+Trabectedin",
                       ident.2 = "2 - oHSV",
                       features = hsv_genes,
                       logfc.threshold = 0,
                       min.pct = 0,
                       only.pos = F)
hsv_DEG

# Save
write.csv(hsv_DEG,
          file = "Data/FigureS4A_A673_hsv_gene_exp.csv")
```

# Antiviral Analysis

## Gene List

## Differential Gene Expression and Gene Set Enrichment Analysis
*Figure 2G*,
*Figure S4B*

Differential gene expression in combination compared to oHSV alone (gene list shown in the table below). GO and KEGG analyses are performed.

No logfc cutoff for DEG & KEGG

```{r deg-trt-0, fig.height=10, fig.width=10, dependson = "hsv_timepoints"}
Idents(A673_hsv_comb_even) <- A673_hsv_comb_even$treatment

# Find DEG between oHSV and combination groups
## Arrange in descending order of avg_log2FC
trtdiff <- FindMarkers(A673_hsv_comb_even,
                       ident.1 = "4 - oHSV+Trabectedin",
                       ident.2 = "2 - oHSV",
                       logfc.threshold = 0,
                       only.pos = F)

trtdiff <- as.data.frame(trtdiff)
trtdiff <- arrange(trtdiff, desc(avg_log2FC))

# Save
write.csv(trtdiff,
          file = "Data/A673_comb_trtdiff_nocutoff.csv")

# Get into correct format for KEGG
degs <- as.vector(trtdiff$avg_log2FC)
names(degs) <- rownames(trtdiff) %>%
  str_replace("HSV1-","")

conversion <-
  clusterProfiler::bitr(geneID = names(degs),
                        fromType = "SYMBOL",
                        toType = "ENTREZID",
                        OrgDb = "org.Hs.eg.db")

keggdegs <-
  as.data.frame(degs) %>%
  rownames_to_column("SYMBOL") %>%
  inner_join(conversion, by = "SYMBOL") %>%
  pull(degs, name = ENTREZID)

gsea_KEGG <- clusterProfiler::gseKEGG(keggdegs,
                                      organism = "hsa",
                                      eps = 0,
                                      pvalueCutoff = 1)

gsea_KEGG <- mutate(gsea_KEGG,
                    p.adjust = -log10(p.adjust))

plots <- list()

if (nrow(gsea_KEGG) > 0) {
  pdf("Plots/A673_comb/dotplot_kegg_nocutoff.pdf",
      height = 8,
      width = 8)
  print(enrichplot::dotplot(gsea_KEGG,
                            x = "NES",
                            showCategory = 30,
                            font.size = 8,
                            color = "NES") +
          labs(title = "Enrichment of Pathways in Combination Therapy",
               subtitle = "KEGG Gene Sets from A673 (No logFC cutoff)")
  )
  dev.off()
  
  # view in rmd
  print(enrichplot::dotplot(gsea_KEGG,
                            x = "NES",
                            showCategory = 100,
                            font.size = 8,
                            color = "NES") +
          labs(title = "Enrichment of Pathways in Combination Therapy",
               subtitle = "KEGG Gene Sets from A673 (No logFC cutoff)")
  )

} else {
  print(ggplot(tibble(x = "A", y = "A",
                      text = "No significant hits"),
               aes(x = x, y = y, label = text)) +
          geom_text() +
          theme(axis.text = element_text(size = 5)) +
          labs(title = "Enrichment of Pathways in Combination Therapy",
               subtitle = "KEGG Gene Sets from A673 (No logFC cutoff)")
  )
}

#### Figure 2G ####
# Look at GSEA enrichment for HSV pathway (hsa05168)
enrichplot::gseaplot(gsea_KEGG,
                     by = "all",
                     title = "Herpes simplex virus 1 infection",
                     geneSetID = "hsa05168")
enrichplot::gseaplot2(gsea_KEGG,
                     title = "Herpes simplex virus 1 infection",
                     geneSetID = "hsa05168",
                     pvalue_table = T)
# save as pdf
pdf(file = "Plots/A673_comb/Figure2G_gsea_curve_hsv_nocutoff.pdf",
    height = 10,
    width = 5)
enrichplot::gseaplot(gsea_KEGG,
                     by = "all",
                     title = "Herpes simplex virus 1 infection",
                     geneSetID = "hsa05168")
enrichplot::gseaplot2(gsea_KEGG,
                     title = "Herpes simplex virus 1 infection",
                     geneSetID = "hsa05168",
                     pvalue_table = T)
dev.off()



# Map onto pathway
gene <- names(keggdegs)[abs(keggdegs) > 0.25]
gsea_enKEGG <- clusterProfiler::enrichKEGG(gene,
                                           organism = "hsa")
## Pull pathway to local
browser <- clusterProfiler::browseKEGG(gsea_KEGG, 'hsa05168')

## Plot expression onto pathway
library("pathview")
library("DOSE")
hsa05168 <- pathview(gene.data = keggdegs,
                     pathway.id = "hsa05168",
                     species = "hsa",
                     limit = list(gene=max(abs(keggdegs)), cpd=1))

## convert gene ID to Symbol
gsea_KEGG_symbol <- setReadable(gsea_KEGG, 'org.Hs.eg.db', 'ENTREZID')
## heatmap
heatplot(gsea_KEGG_symbol, foldChange=keggdegs, showCategory=30)

gsea_KEGG_symbol2 <- pairwise_termsim(gsea_KEGG_symbol)
treeplot(gsea_KEGG_symbol2, color = "NES")

# save KEGG pathway object
save(gsea_KEGG,
     file = "Data/gsea_KEGG__A673_nocutoff.RData")
```

```{r wholehsvpath_exp, dependson = "deg-trt-0"}
# Generate list of all genes in the HSV1 pathway plot (hsa05168)
hsa05168_genelist <- hsa05168$plot.data.gene[["labels"]]

# Only keep genes found in the A673 data
hsa05168_geneintersect <- intersect(hsa05168_genelist,
                                    rownames(A673_hsv_comb_even))

# Calculate fold change/p values for genes from the HSV1 pathway in the A673 data
hsa05168_DEG <- FindMarkers(A673_hsv_comb_even,
                            ident.1 = "4 - oHSV+Trabectedin",
                            ident.2 = "2 - oHSV",
                            features = hsa05168_geneintersect,
                            logfc.threshold = 0,
                            min.pct = 0,
                            only.pos = F)

# Order by absolute value of the log2 fold change (decreasing)
hsa05168_DEG <- hsa05168_DEG[order(-abs(hsa05168_DEG$avg_log2FC)),]

write.csv(hsa05168_DEG,
          file = "Data/A673_hsv_comb_even_hsa05168_DEG.csv")
```

### Table of HSV1 pathway DEGs

```{r trtdiff, dependson = "deg-trt-0"}
# Create interactive datatable showing DEG list
#trtdiff <- as.data.table(trtdiff)
DT::datatable(hsa05168_DEG, 
              rownames = TRUE,
              extensions = c('FixedColumns',
                             'Buttons'),
              options = list(
                pageLength = 10,
                scrollX = TRUE,
                scrollCollapse = TRUE,
                dom = 'Bfrtip',
                buttons = c('copy',
                            'csv',
                            'excel')
              ))
dev.off()
```

#### Fold Changes for Genes from HSV1 KEGG Pathway
*Figure S4B*

Calculated for oHSV+Trab compared to oHSV alone. 

```{r antiviraltbl, dependson = "deg-trt-0", fig.width=20}
backconversion <-
  clusterProfiler::bitr(geneID = gsea_KEGG[["hsa05168"]],
                        fromType = "ENTREZID",
                        toType = "SYMBOL",
                        OrgDb = "org.Hs.eg.db")

Idents(A673_hsv_comb_even) <- A673_hsv_comb_even$treatment
antiviralFC <- FindMarkers(A673_hsv_comb_even,
                           features = backconversion[["SYMBOL"]],
                           ident.1 = "4 - oHSV+Trabectedin",
                           ident.2 = "2 - oHSV",
                           logfc.threshold = 0) 
antiviralFC <- antiviralFC[order(-abs(antiviralFC$avg_log2FC)),]

write.csv(antiviralFC,
          file = "Data/FigureS4B_A673_hsv_comb_even_antiviralFC.csv")

# Create interactive datatable showing DEG list
DT::datatable(antiviralFC[order(antiviralFC$avg_log2FC),], 
              rownames = TRUE,
              extensions = c('FixedColumns',
                             'Buttons'),
              options = list(
                pageLength = 10,
                scrollX = TRUE,
                scrollCollapse = TRUE,
                dom = 'Bfrtip',
                buttons = c('copy',
                            'csv',
                            'excel')
              ))
```

# EWS-FLI targets
Does trabectedin impact EWS-FLI related gene expression?
```{r ewsfli, dependson = "downsample"}
# upreg by ewsfli
VlnPlot(A673_hsv_comb_even,
        features = c("NR0B1","GLI1","NKX2-2","PTCH","GAS1","CDK2","FOXM1"),
        group.by="treatment",
        split.by = "treatment",
        log = T)

# down reg. by ewsfli
VlnPlot(A673_hsv_comb_even,
        features = c("FOXO1","FOXO3A","FOXO4","FOXO6","LOX","LOXL1","LOXL4"),
        group.by="treatment",
        split.by = "treatment",
        log = T)

Idents(A673_hsv_comb_even) <- A673_hsv_comb_even$treatment
FoldChange(A673_hsv_comb_even,
           ident.1 = "3 - Trabectedin",
           ident.2 = "1 - Control",
           features = c("NR0B1","GLI1","NKX2-2","PTCH1","GAS1","CDK2","FOXM1",
                        "FOXO1","FOXO3","FOXO4","FOXO6","LOX","LOXL1","LOXL4"))
```


```{r sessioninfo}
sessionInfo()
```
