---
title: "Repeat K7M2 scRNAseq with HSV exon reference"
author: "Emily Franz"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 5
    number_sections: false
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  echo = TRUE,
  cache = TRUE,
  collapse = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  cache.lazy = FALSE)
```

```{r lib, cache = FALSE}
library(Seurat)
library(tidyverse)
library(ggsci)
library(rrrSingleCellUtils)
library(ggplot2)
library(patchwork)
source("scSeurat-addhsv.R")
# Set the random generator seed so that results are reproducible.
set.seed(888)

# Create files for saving if do not exist
if(!file.exists("Data")){
  dir.create("Data")
}
if(!file.exists("Plots")){
  dir.create("Plots")
}
if(!file.exists("Plots/K7M2_2")){
  dir.create("Plots/K7M2_2")
}
```

# Load and Process

## Tumor Samples - Mouse
Load all the different K7M2 datasets, note that HSV genes included but 
filter out human. (K7M2 is a mouse osteosarcoma model.)

```{r define_samples}
sample_names <- c('S0238_hsv',
                  'S0239_hsv',
                  'S0240_hsv',
                  'S0241_hsv')

sample_labels <- c("1 - Control",
                   "2 - oHSV",
                   "3 - Trabectedin",
                   "4 - oHSV+Trabectedin")
# correlate sample names and sample labels
names(sample_labels) <- sample_names

sample_titles <- c("Control",
                   "oHSV",
                   "Trabectedin",
                   "Combination")
names(sample_titles) <- sample_names
```

```{r load, fig.height=5, fig.width=5, dependson="define_samples"}
#Load in objects and determine cutoffs via plots
qcplots <- list()
for (item in sample_names) {
  obj <- tenx_load_qc(paste("/gpfs0/home2/gdrobertslab/lab/Counts/",
                            item,
                            "/filtered_feature_bc_matrix/",
                            sep = ""),
                      mouse_hsv = TRUE)

  plot1 <- FeatureScatter(obj,
                          feature1 = "nCount_RNA",
                          feature2 = "percent.mt") +
    theme(legend.position="none") +
    theme(panel.grid = element_line(colour = "grey80",
                                    linetype = "solid"))
  
  plot2 <- FeatureScatter(obj,
                          feature1 = "nCount_RNA",
                          feature2 = "nFeature_RNA")  +
    theme(legend.position="none") +
    theme(panel.grid = element_line(colour = "grey80",
                                    linetype = "solid"))
  
  qcplots[[sample_titles[[item]]]] <- plot1 + plot2 +
    plot_annotation(title = sample_titles[[item]])
  
  assign(sample_titles[[item]], obj) #makes each sample an individual Seurat obj without overwriting
}
```

Look at HSV transcripts vs mt transcripts and HSV transcripts vs nCounts.

oHSV-treated sample:

```{r scatter_hsv, fig.height=5, fig.width=10, dependson=c("define_samples","load")}
# Check for hsv presence by plotting percent hsv genes vs ncount
plot1_hsv <- FeatureScatter(oHSV,
                            feature1 = "nCount_RNA",
                            feature2 = "percent.hsv") +
  theme(legend.position="none") +
  theme(axis.line = element_line(colour = "grey80"))

plot2_hsv <- FeatureScatter(oHSV,
                            feature1 = "percent.mt",
                            feature2 = "percent.hsv") +
  theme(legend.position="none") 

plot1_hsv +
  plot2_hsv +
  plot_annotation(title = 'oHSV-treated Tumor')
```

Combination-treated sample:

```{r scatter_combo, fig.height=5, fig.width=10, dependson=c("define_samples","load")}
# Check for hsv presence by plotting percent hsv genes vs ncount
plot1_hsv <- FeatureScatter(Combination,
                            feature1 = "nCount_RNA",
                            feature2 = "percent.hsv") +
  theme(legend.position="none")

plot2_hsv <- FeatureScatter(Combination,
                            feature1 = "percent.mt",
                            feature2 = "percent.hsv") +
  theme(legend.position="none")

plot1_hsv +
  plot2_hsv +
  plot_annotation(title = 'Combination-treated Tumor')
```

Using the previous violin plots, choose cutoffs to proceed with for processing.

```{r load_final, dependson=c("define_samples","load")}
#### QC Subset for Loaded Samples ####
# set max ncounts 
nCount_RNA_max <- list(55000,
                       100000,
                       75000,
                       100000)
names(nCount_RNA_max) <- sample_names

# set min for ncounts 
nCount_RNA_min <- c(800,
                    800,
                    800,
                    800)
names(nCount_RNA_min) <- sample_names

# set max nfeatures 
nFeature_RNA_max <- list(6500,
                         7700,
                         7000,
                         7500)
names(nFeature_RNA_max) <- sample_names

# set % mt genes 
mt_max <- c(20,
            20,
            20,
            20)
names(mt_max) <- sample_names

# Do not need to plot violin plot since completed this previously to determine the cutoffs used here.
for (item in sample_names) {
  obj <- tenx_load_qc(paste("/gpfs0/home2/gdrobertslab/lab/Counts/",
                            item,
                            "/filtered_feature_bc_matrix/",
                            sep = ""),
                      mouse_hsv = TRUE,
                      violin_plot = FALSE)
  
  obj <- subset(obj,
                subset = 
                  nCount_RNA < nCount_RNA_max[[item]] &
                  nCount_RNA > nCount_RNA_min[[item]] &
                  nFeature_RNA < nFeature_RNA_max[[item]] &
                  percent.mt < mt_max[[item]])
  
  obj$treatment <- sample_labels[[item]]
  
  # Make each sample an individual Seurat obj without overwriting
  assign(sample_titles[[item]], obj)
}
```

## Downsample

```{r downsample, fig.height=3, fig.width=3, dependson=c("define_samples","load","load-final")}
#### Merge ####
# Merge cells from the different treatment groups into one Seurat object
K7M2_hsv <- 
  merge(Control,
        y = c(oHSV,
              Trabectedin,
              Combination),
        add.cell.ids = c("Control",
                         "oHSV",
                         "Trabectedin",
                         "Combination"),
        project = "Trabectedin plus HSV1716 in Low Passage K7M2")

#### Downsample ####
# Set seed before random downsampling
set.seed(132)

# Label cells by treatment group
Idents(K7M2_hsv) <- K7M2_hsv$treatment

# Subset the highest number that makes all treatment groups equal in cell number
K7M2_hsv_even <-
  subset(K7M2_hsv,
         downsample = table(K7M2_hsv$treatment) %>%
           min())

## Check that worked: cell numbers across treatments should be equal
table(K7M2_hsv_even$treatment)
```

## Normalize and Process

```{r process, fig.height=4, fig.width=4, dependson=c("define_samples","load","load-final","downsample")}
#### Normalize, scale, process, cluster ####
K7M2_hsv_even <-
  NormalizeData(K7M2_hsv_even) %>%
  FindVariableFeatures() %>%
  ScaleData(features = rownames(K7M2_hsv_even))

#### Variable Features ####
K7M2_hsv_even <-
  RunPCA(K7M2_hsv_even,
         features = VariableFeatures(object = K7M2_hsv_even))

# Use Elbow Plots to determine the PCs to use
ElbowPlot(K7M2_hsv_even,
          ndims = 30) + 
  theme(panel.grid = element_line(color = "#8ccde3",
                                  linetype = 2))
```

### Clustering Resolution {.tabset}

#### Res0.1
```{r umap, fig.height=5, fig.width=10, dependson=c("define_samples","load","load-final","process")}
K7M2_hsv_even <-
  FindNeighbors(K7M2_hsv_even, dims = 1:17) %>%
  FindClusters(resolution = 0.1) %>%
  RunUMAP(dims = 1:17)

DimPlot(K7M2_hsv_even,
        reduction = "umap",
        split.by = "treatment",
        pt.size = 1,
        label = T) +
  coord_fixed()
```

#### Res0.2
```{r umap2, fig.height=5, fig.width=10, dependson=c("define_samples","load","load-final","downsample","process")}
K7M2_hsv_even <-
  FindClusters(K7M2_hsv_even,
               resolution = 0.2)

DimPlot(K7M2_hsv_even,
        reduction = "umap",
        split.by = "treatment",
        pt.size = 1,
        label = T) +
  coord_fixed()
```

#### Res0.3
```{r umap3, fig.height=5, fig.width=10, dependson=c("define_samples","load","load-final","downsample","process")}
K7M2_hsv_even <-
  FindClusters(K7M2_hsv_even,
               resolution = 0.3)

DimPlot(K7M2_hsv_even,
        reduction = "umap",
        split.by = "treatment",
        pt.size = 1,
        label = T) +
  coord_fixed()
```

#### Res0.5
```{r umap5, fig.height=5, fig.width=10, dependson=c("define_samples","load","load-final","downsample","process")}
K7M2_hsv_even <-
  FindClusters(K7M2_hsv_even,
               resolution = 0.5)

DimPlot(K7M2_hsv_even,
        reduction = "umap",
        split.by = "treatment",
        pt.size = 1,
        label = T) +
  coord_fixed()
```

#### Res0.7
```{r umap7, fig.height=5, fig.width=10, dependson=c("define_samples","load","load-final","downsample","process")}
K7M2_hsv_even <-
  FindClusters(K7M2_hsv_even,
               resolution = 0.7)

DimPlot(K7M2_hsv_even,
        reduction = "umap",
        split.by = "treatment",
        pt.size = 1,
        label = T) +
  coord_fixed()
```

```{r orig_dim, dependson=c("define_samples","load","load-final","downsample","process")}
DimPlot(K7M2_hsv_even,
        split.by = "treatment") +
  coord_fixed()

# Save object prior to recursive clustering
save(K7M2_hsv_even,
     file = "Data/K7M2_hsv_even_rep2.RData")
```

# Use Recursive Clustering Instead

```{r recurl, dependson=c("define_samples","load","load-final","downsample","process","umap","umap5","downsample")}
# Load the function code (developed by Matthew Cannon)
source("/gpfs0/home2/gdrobertslab/lab/Analysis/Matt/10xRNAatac/test1/recurl.R")

# Rename res metadata so does not get overwritten
K7M2_hsv_even[["res0.1"]] <- K7M2_hsv_even$RNA_snn_res.0.1
K7M2_hsv_even[["res0.3"]] <- K7M2_hsv_even$RNA_snn_res.0.3
K7M2_hsv_even[["res0.5"]] <- K7M2_hsv_even$RNA_snn_res.0.5

# Recluster using recursive clustering and plot
krecurlustered <-
    recurluster(K7M2_hsv_even,
                do_plots = TRUE,
                parallel = FALSE)

krecurlustered <- 
  ScaleData(krecurlustered) %>%
  FindVariableFeatures() %>%
  RunPCA() %>%
  FindNeighbors(dims = 1:17) %>%
  RunUMAP(dims = 1:17)

save(krecurlustered,
     file = "Data/K7M2_hsv_even_rep2_recurl.RData")
```

```{r recurl-plot, fig.height=5, fig.width=10, dependson=c("define_samples","load","load-final","downsample","process","umap","umap5","downsample","recurl")}
library(Polychrome)
# Create color palette
par(mfrow=c(2,3),
    mar=c(0.5,0.5,0.5,0.5))
C50_29 = createPalette(50,
                       c("#ff0000", "#ffff00", "#00ff00", "#0000ff"),
                       range = c(20, 90))

pdf("Plots/krecurlustered_level2.pdf",
    height = 10,
    width = 12)
DimPlot(krecurlustered,
        group.by="clust_2",
        cols = unname(C50_29)) +
  coord_fixed() +
  ggtitle("Recursive Clustering Assignment")
dev.off()

DimPlot(krecurlustered,
        cols = unname(C50_29))

DimPlot(krecurlustered,
        cols = unname(C50_29),
        group.by = "clust_2",
        #label = TRUE,
        repel = TRUE,
        split.by = "treatment")
```

```{r recurl-deg, dependson=c("define_samples","load","load-final","downsample","process","umap","umap5","downsample","recurl")}
Idents(krecurlustered) <- krecurlustered$clust_2
krecurlustered_markers <- FindAllMarkers(krecurlustered,
                                         min.pct = 0.25,
                                         logfc.threshold = 0.25,
                                         only.pos = F)

# Hard to work with this function/datatable so this step removes genes not in the top 10 DEG for all clusters 
# Print top 10 in each cluster
krecurlustered_markers %>%
  group_by(cluster) %>%
  top_n(n = 10, wt = avg_log2FC) -> t

# Get dataframe into proper format 
t <- as.data.frame(t)
#t$cluster <- as.numeric(as.character(t$cluster))
#t <- t[order(t$cluster),]
t <- t[,c(7, 1:6)]
t <- t[,c(7, 1:6)]
#could also use arrange() to sort by column

# Create interactive datatable showing top 10 DEG in each cluster 
DT::datatable(t, 
              rownames = FALSE,
              extensions = c('FixedColumns',
                             'Buttons'),
              options = list(
                pageLength = 10,
                scrollX = TRUE,
                scrollCollapse = TRUE,
                dom = 'Bfrtip',
                buttons = c('copy',
                            'csv',
                            'excel')
              ))
```

# Assign Cell Identities

## SingleR Cell Predictions

This is an automatic prediction method for cell types. Here, I specifically use the SingleR mouse immune reference for predicted cell assignment of cell clusters. Often this method is a general assignment and may miss more specific cell subtypes, especially for macrophages.

```{r singlr, dependson=c("define_samples","load","load-final","downsample","process","umap","umap5","downsample")}
# Automatic prediction of cell types using SingleR
# BiocManager::install("SingleCellExperiment")
library(SingleR)
library(celldex)
library(SingleCellExperiment)

# Convert the Seurat object to a SCE object
krecurlustered_sce <- as.SingleCellExperiment(krecurlustered)

# Load the mouse reference library & immune-specific mouse data
ref1 <- MouseRNAseqData()
ref2 <- ImmGenData()

# Make cell type predictions using SingleR
Tumor_hsv_pred_types <- SingleR(test = krecurlustered_sce,
                                ref = list(ref1,
                                           ref2),
                                labels = list(ref1$label.main,
                                              ref2$label.main))
# Clean up environment
rm(krecurlustered_sce, ref1, ref2)
```

```{r singler_cont, dependson=c("define_samples","load","load-final","downsample","process","umap","umap5","downsample","singlr")}
# Transfer labels back to the Seurat object and plot
krecurlustered$singleR2 <- Tumor_hsv_pred_types$labels

maxv <- apply(Tumor_hsv_pred_types$scores, 
              MARGIN = 1, 
              function(x) max(x,
                              na.rm = TRUE))
hist(maxv,
     n = 200)

krecurlustered$singleRscore <- maxv  #scores may be a matrix

as.data.frame(krecurlustered@meta.data) %>%
  as_tibble() %>%
  ggplot(aes(x = singleRscore)) +
  geom_histogram(bins = 200) +
  facet_wrap(~singleR2,
             ncol = 3,
             scales = "free_y") #****
  
table(krecurlustered$singleR2,
      krecurlustered$treatment)
```

```{r singlr_plot, dependson=c("define_samples","load","load-final","downsample","process","umap","umap5","downsample","singlr")}
Idents(krecurlustered) <- krecurlustered$singleR2 
# Plot
krecurlustered <- 
  RenameIdents(krecurlustered,
               "Dendritic cells" = "DC") %>%
  StashIdent(save.name = "SingleR2")

DimPlot(krecurlustered,
        reduction = "umap",
        pt.size = 0.5,
        label = T,
        repel = T) +
  coord_fixed() +
  ggtitle("Mouse Reference")

pdf("Plots/K7M2_2/singlr.pdf",
    height = 10,
    width = 10)
r_dim_plot(krecurlustered,
           reduction = "umap",
           pt.size = 0.5,
           label = T,
           repel = T) +
  coord_fixed() +
  ggplot2::theme(legend.position = "right") +
  ggtitle("Mouse Reference")
dev.off()
```

```{r singlr_plot2, fig.width=12, fig.height=5, dependson=c("define_samples","load","load-final","downsample","process","umap","umap5","downsample","singlr")}
# Rename cell ids
krecurlustered <- 
  RenameIdents(krecurlustered,
               "Dendritic cells" = "DC",
               "Tgd" = "T cells",
               "Stem cells" = "Tumor cells",
               "Granulocytes" = "Neutrophils",
               "Erythrocytes" = "Tumor cells",
               "Adipocytes" = "Tumor cells",
               "Fibroblasts" = "Tumor cells") %>%
  StashIdent(save.name = "SingleR2ed")

k7m2_ids <- r_dim_plot(krecurlustered,
                       reduction = "umap",
                       pt.size = 0.5,
                       label = T,
                       repel = T,
                       split.by = "treatment") +
  coord_fixed() +
  ggtitle("Mouse Reference")
k7m2_ids

pdf("Plots/K7M2_2/k7m2_renamesinglr_split.pdf",
    height = 10,
    width = 20)
k7m2_ids
dev.off()
```

### Label cell identities

```{r cellname}
krecurlustered$cell_ids <-
  str_replace_all(krecurlustered$clust_3,
                  c(
                    "clust_0.0.." = "Tumor cells",
                    "clust_0.1.." = "Tumor cells",
                    "clust_0.2.." = "Tumor cells",
                    "clust_0.3.." = "Tumor cells",
                    "clust_1.1.." = "DC",
                    "clust_1.2.." = "DC",
                    "clust_2.3.." = "NK cells",
                    "clust_2.0.." = "NK cells",
                    "clust_2.2.." = "CD8 T cells",
                    "clust_2.4.." = "Prolif NK and T cells",
                    "clust_2.1.." = "Activated NK and CD8 T cells",
                    "clust_2.5.." = "Tregs",
                    "clust_2.6.." = "CD4 T cells",
                    "clust_3.1.." = "Endothelial cells",
                    "clust_3.0.." = "Fibroblasts",
                    "clust_5.1.." = "Neutrophils",
                    "clust_5.0.." = "Neutrophils",
                    "clust_4.0.." = "Mast cells",
                    "clust_4.1.." = "Mast cells",
                    "clust_1.0.0" = "Monocytes",
                    "clust_1.0.1" = "Monocytes",
                    "clust_1.0.2" = "Macrophages"
                  )) 

labeled_dim <- r_dim_plot(krecurlustered,
                          group.by = "cell_ids",
                          label = F,
                          shuffle = T,
                          #repel = T,
                          split.by = "treatment") +
  coord_fixed() + 
  ggplot2::theme(legend.position = "right")

pdf(file = "Plots/K7M2_2/krecurl_cell_ids_dim_2.pdf",
    height = 5,
    width = 20)
labeled_dim
dev.off()

r_dim_plot(krecurlustered,
           split.by = "treatment",
           group.by = "cell_ids") +
  coord_fixed()
```

```{r anysubset2}
library(dittoSeq)
# set idents 
Idents(krecurlustered) <- krecurlustered$cell_ids
# subset cells
tumor <- subset(krecurlustered,
                idents = "Tumor cells")
nk_t <- subset(krecurlustered,
               idents = c("NK cells",
                          "CD8 T cells",
                          "Prolif NK and T cells",
                          "Activated NK and CD8 T cells",
                          "Tregs",
                          "CD4 T cells"))
myeloid <- subset(krecurlustered,
              idents = c("DC",
                         "Monocytes",
                         "Macrophages"))

subsets <- list(Tumor = tumor,
                NK_T = nk_t,
                Myeloid = myeloid)

for (celltype in names(subsets)) {
  # Set seed before random downsampling
  set.seed(132)
  # Print out the identifying cell label
  print(celltype)
  # show relative numbers of cell ids per treatment in each subset
  table(subsets[[celltype]]@meta.data[["cell_ids"]],
        subsets[[celltype]]@meta.data[["treatment"]]) %>%
    print()
  # Prepare for downsampling
  Idents(subsets[[celltype]]) <- subsets[[celltype]]$treatment
  # Display overall cell count prior to downsampling
  table(subsets[[celltype]]$treatment) %>%
    print()
  # show cell quantities
  print(dittoBarPlot(subsets[[celltype]],
                     var = "cell_ids",
                     group.by = "treatment")) +
    ggtitle(celltype)
  # Subset the highest number that makes all treatment groups equal in cell number
  type <-
    subset(subsets[[celltype]],
           downsample = table(subsets[[celltype]]$treatment) %>%
             min())
  ## Check that worked: cell numbers across treatments should be equal
  print(paste("Downsampled",
              celltype))
  print(table(type$treatment))
  # Assign to an object
  assign(paste0(celltype,
                "_downsampled"),
         type)
  # show cell quantities
  print(dittoBarPlot(subsets[[celltype]],
                     var = "cell_ids",
                     group.by = "treatment",
                     data.out = TRUE))
  pdf(paste0("Plots/K7M2_2/",
             celltype,
             "cellgroup_subsets.pdf"))
  print(dittoBarPlot(subsets[[celltype]],
                     var = "cell_ids",
                     group.by = "treatment"))
  dev.off()
  ## Plot
  print(
    DimPlot(type,
            split.by = "treatment") +
    ggtitle(paste(celltype, "Cells"))
  )
}
```

### Analyze T and NK Cytotoxicity Expression

```{r nktsubsets}
# set idents 
Idents(krecurlustered) <- krecurlustered$cell_ids
# subset cells
nk <- subset(krecurlustered,
             idents = "NK cells")
cd8t <- subset(krecurlustered,
               idents = "CD8 T cells")
activ_nkt <- subset(krecurlustered,
                    idents = "Activated NK and CD8 T cells")
prolif_nkt <- subset(krecurlustered,
                    idents = "Prolif NK and T cells")

subsets <- list(NK = nk,
                CD8_T = cd8t,
                Activated_NK_T = activ_nkt,
                Prolif_NK_T = prolif_nkt)

for (celltype in names(subsets)) {
  # Set seed before random downsampling
  set.seed(132)
  Idents(subsets[[celltype]]) <- subsets[[celltype]]$treatment
  print(celltype)
  print(table(subsets[[celltype]]$treatment))
  # Subset the highest number that makes all treatment groups equal in cell number
  type <-
    subset(subsets[[celltype]],
           downsample = table(subsets[[celltype]]$treatment) %>%
             min())
  ## Check that worked: cell numbers across treatments should be equal
  print(paste("Downsampled",
              celltype))
  print(table(type$treatment))
  # Label cells by treatment group
  Idents(type) <- type$treatment
  # Assign to an object
  assign(celltype, type)
  ## Plot
  print(DimPlot(type,
                split.by = "treatment"))
  ## Plot granzyme expression
  pdf(paste0("Plots/K7M2_2/",
             celltype,
             "_GzmVln",
             ".pdf"),
      height = 14,
      width = 10)
  print(VlnPlot(type,
                features = c("Gzmd",
                             "Gzme",
                             "Gzmc",
                             "Gzmf",
                             "Gzmg",
                             "Gzma",
                             "Gzmb",
                             "Prf1",
                             "Xcl1"),
                log = T,
                split.by = "treatment",
                group.by = "treatment",
                cols = c("grey", "darkblue", "orangered", "purple"))
        )
  dev.off()
}
```

Find and plot immune genes of interest 

```{r granzymes, fig.height = 11,fig.width=9,dependson="anysubset2"}
#Granzymes
Idents(nk_t) <- nk_t$treatment
nkt_markers <- FindMarkers(nk_t,
                           ident.1 = "4 - oHSV+Trabectedin",
                           ident.2 = "2 - oHSV",
                           min.pct = 0,
                           logfc.threshold = 0,
                           features = c("Gzmd",
                                        "Gzme",
                                        "Gzmc",
                                        "Gzmf",
                                        "Gzmg",
                                        "Gzma",
                                        "Gzmb",
                                        "Prf1",
                                        "Xcl1"),
                           only.pos = T) %>%
  arrange(-abs(avg_log2FC)) %>%
  as.data.frame()

# Create interactive datatable showing top 10 DEG in each cluster 
DT::datatable(nkt_markers, 
              rownames = FALSE,
              extensions = c('FixedColumns',
                             'Buttons'),
              options = list(
                pageLength = 15,
                scrollX = TRUE,
                scrollCollapse = TRUE,
                dom = 'Bfrtip',
                buttons = c('copy',
                            'csv',
                            'excel')
              ))

# Plot genes of interest
VlnPlot(nk_t,
        features = c("Gzmd",
                     "Gzme",
                     "Gzmc",
                     "Gzmf",
                     "Gzmd",
                     "Gzma",
                     "Gzmb",
                     "Prf1",
                     "Xcl1"),
        log = T,
        split.by = "treatment",
        cols = c("grey", "darkblue", "orangered", "purple"))

pdf("Plots/K7M2_2/nk_t-gzms.pdf",
    height = 14,
    width = 10)
VlnPlot(nk_t,
        features = c("Gzmd",
                     "Gzme",
                     "Gzmc",
                     "Gzmf",
                     "Gzmg",
                     "Gzma",
                     "Gzmb",
                     "Prf1",
                     "Xcl1"),
        log = T,
        split.by = "treatment",
        cols = c("grey", "darkblue", "orangered", "purple"))
dev.off()

# Check fold change differences in trab. vs control
FoldChange(nk_t,
           ident.1 = "3 - Trabectedin",
           ident.2 = "1 - Control",
           features = c("Gzmd",
                        "Gzme",
                        "Gzmc",
                        "Gzmf",
                        "Gzmg",
                        "Gzma",
                        "Gzmb",
                        "Prf1",
                        "Xcl1")) %>%
  print()
```

### Death receptor signaling

```{r fig.height=4, fig.width=9}
# Death receptor signaling
Idents(tumor) <- tumor$treatment
VlnPlot(tumor,
        features = c("Fas",
                     "Tnfrsf10b",
                     "Hspa1a"),
        log = T,
        split.by = "treatment")
VlnPlot(nk_t,
        features = c("Fasl",
                     "Tnfsf10",
                     "Tbx21"),
        log = T,
        split.by = "treatment")

pdf("Plots/K7M2_2/deathreceptors-tumor.pdf",
    height = 5,
    width = 10)
VlnPlot(tumor,
        features = c("Fas",
                     "Tnfrsf10b",
                     "Hspa1a"),
        log = T,
        split.by = "treatment")
dev.off()

pdf("Plots/K7M2_2/deathsignals-nkt.pdf",
    height = 5,
    width = 10)
VlnPlot(nk_t,
        features = c("Fasl",
                     "Tnfsf10",
                     "Tbx21"),
        log = T,
        split.by = "treatment")
dev.off()
```

# Subset Cell ID Analysis

Uses krecurlustered subsets

```{r cell_subset, fig.width=10, fig.height=10}
library(cowplot)
library(forcats)
library(ggstance)
source("enrichplot_codeforbarplot.R")
library(enrichplot)
library(DOSE)
#source("dgea.R")

# Change idents to cell identities
Idents(krecurlustered) <- krecurlustered$cell_ids
# Create list for plots
keggplots <- list()
pathwayenrich <- list()
# match color to what is used in r_dimplot (above)
cols2id <- c("#D43F3AFF", "#EEA236FF", "#357EBDFF", "#5CB85CFF",
             "#B8B8B8FF", "#9632B8FF", "#46B8DAFF", "#90302DFF",
             "#A66D04FF", "#2D577FFF", "#3E7E3EFF", "#7D7D7DFF",
             "#6D1D87FF", "#097F9AFF", "#FF6E6AFF", "#FFBB70FF",
             "#68A4E3FF", "#79D379FF", "#CDCDCDFF", "#BF6BE2FF",
             "#69D1F3FF")
names(cols2id) <- 
  sort(levels(krecurlustered))

# DEG and KEGG pathway analyses on subsetted cell types
for (cell_id in levels(krecurlustered)) {
  # subset each cell type
  cell_type <- subset(krecurlustered,
                      idents = cell_id)
  # plot by treatment group (for figures)
  cell_type_plot <-
    DimPlot(cell_type,
            group.by = "cell_ids",
            split.by = "treatment",
            cols = scales::alpha(c(cols2id[[cell_id]],
                                   sample(rainbow(1000))),
                                 0.6)) +
    xlim(-10,10) + 
    ylim(-10,10) +
    coord_fixed() +
    ggtitle(paste(cell_id, "prior to downsampling"))
  ## print in rmd output
  print(cell_type_plot)
  ## create pdf for use in figures
  pdf(file = paste0("Plots/K7M2_2/",
                    cell_id,
                    "fromkrecurl_dimplot.pdf"),
      height = 4,
      width = 16)
  print(cell_type_plot)
  dev.off()
  # create table of cell type # per treatment group
  Idents(cell_type) <- cell_type$treatment
  print(table(cell_type$treatment))
  # randomly downsample to have the same number of cells per treatment group
  cell_type <- subset(cell_type,
                      downsample = table(cell_type$treatment) %>%
                        min())
  # provide status message on which cell type is being analyzed
  print(paste("Analyzing Downsampled", cell_id))
  # check that cell numbers are now equal between treatment groups
  print(table(cell_type$treatment))
  # if more then 20 cells per group, perform DEG and pathway analyses
  if (min(table(cell_type$treatment)) > 20) {
    Idents(cell_type) <- cell_type$treatment

    #### KEGG analysis ####
    # find DEGs with |logfc| > 0.05 and arrange in decreasing order
    trtdiff <- FindMarkers(cell_type,
                           ident.1 = "4 - oHSV+Trabectedin",
                           ident.2 = "2 - oHSV",
                           logfc.threshold = 0.05,
                           only.pos = F) %>%
      arrange(-abs(avg_log2FC))
    print(head(trtdiff, n=20))
    # Plot top 10 DEGs
    pdf(paste0("Plots/K7M2_2/",
               cell_id,
               "_topdegs.pdf"),
        height = 15,
        width = 15)
    print(VlnPlot(cell_type,
                  features = head(rownames(trtdiff)),
                  group.by = "treatment",
                  cols = c("grey",
                           "darkblue",
                           "orangered",
                           "purple"),
                  log = T)
          )
    dev.off()
    # put DEG output into proper form for kegg analysis
    trtdiff <- as.data.frame(trtdiff)
    trtdiff <- arrange(trtdiff, desc(avg_log2FC))
    degs <- as.vector(trtdiff$avg_log2FC)
    names(degs) <- rownames(trtdiff) %>%
      str_replace("HSV1-","")
    # convert gene names to entrezid
    conversion <-
      clusterProfiler::bitr(geneID = names(degs),
                            fromType = "SYMBOL",
                            toType = "ENTREZID",
                            OrgDb = "org.Mm.eg.db")
    # put proper gene names into correct format to input into kegg analysis
    keggdegs <-
      as.data.frame(degs) %>%
      rownames_to_column("SYMBOL") %>%
      inner_join(conversion,
                 by = "SYMBOL") %>%
      pull(degs,
           name = ENTREZID)
    # perform kegg pathway analysis and visualize
    gsea_KEGG <- clusterProfiler::gseKEGG(keggdegs,
                                          organism = "mmu",
                                          eps = 0)
    gsea_KEGG <- mutate(gsea_KEGG,
                        p.adjust = -log10(p.adjust))
    #### Plot ####
    if (nrow(gsea_KEGG) > 0) {
      # Dotplot
      keggplots[["KEGG_Dot"]][[cell_id]] <- 
        enrichplot::dotplot(gsea_KEGG,
                            x = "NES",
                            showCategory = 25,
                            font.size = 8,
                            color = "NES") +
        labs(title = "Enrichment of Pathways in Combination Therapy",
             subtitle =  paste("KEGG Gene Sets from", cell_id, "in K7M2"))
      ## print in output
      print(keggplots[["KEGG_Dot"]][[cell_id]])
      
      # Bar plot
      keggplots[["KEGG_Bar"]][[cell_id]] <- 
        barplot.enrichResult(gsea_KEGG,
                             font.size=8,
                             label_format=15) +
        labs(title = "Enrichment of Pathways in Combination Therapy",
             subtitle =  paste("KEGG Gene Sets from", cell_id, "in K7M2"))
      ## print in output
      print(keggplots[["KEGG_Bar"]][[cell_id]])
      
      # Bar Plot 2
      keggplots[["KEGG_Bar2"]][[cell_id]] <- arrange(gsea_KEGG,
                                                    abs(NES)) %>% 
        group_by(sign(NES)) %>%
        ggplot(aes(NES,
                   fct_reorder(Description, NES),
                   fill = NES),
               showCategory = 30) + 
        geom_col(orientation='y') + 
        scale_fill_continuous(low="#357EBDFF",
                              high="#D43F3AFF",
                              guide=guide_colorbar(reverse=FALSE)) + 
        theme_minimal() +
        ylab(NULL) +
        ggtitle(paste("Barplot of",
                      cell_id,
                      "KEGG: Version 2"))
      print(keggplots[["KEGG_Bar2"]][[cell_id]])
      
      # heatplot
      gsea_KEGG_symbol <- setReadable(gsea_KEGG,
                                      'org.Mm.eg.db',
                                      'ENTREZID')
      pdf(file = paste0("Plots/K7M2_2/",
                        cell_id,
                        "_keggheatplot.pdf"),
          height = 10,
          width = 20)
      print(heatplot(gsea_KEGG_symbol,
               foldChange = keggdegs,
               showCategory = 30) + 
              scale_fill_gradient2(low="#357EBDFF",
                                   mid = "lightgrey",
                                   high="#D43F3AFF",
                                   na.value = "tan",
                                   midpoint = 0,
                                   guide=guide_colorbar(reverse=FALSE))
            )
      dev.off()
      #save enrichment object
      pathwayenrich[["KEGG"]][[cell_id]] <- gsea_KEGG
    } else {
      # if no signficant pathways are found, state this in (empty) output plot
      nohits <-
        ggplot(tibble(x = "A", y = "A",
                      text = "No significant hits"),
               aes(x = x, y = y, label = text)) +
        geom_text() +
        theme(axis.text = element_text(size = 5)) +
        labs(title = "Enrichment of Pathways in Combination Therapy",
             subtitle = paste("KEGG Gene Sets from", cell_id, "in K7M2"))
      keggplots[["KEGG_Dot"]][[cell_id]] <- nohits
      keggplots[["KEGG_Bar"]][[cell_id]] <- nohits
      keggplots[["KEGG_Bar2"]][[cell_id]] <- nohits
      # print plots in output
      print(keggplots[["KEGG_Dot"]][[cell_id]])
    }

    pdf(paste0("Plots/K7M2_2/",
               cell_id,
               "_KEGGenrichmentdot.pdf"),
        height = 6,
        width = 8)
    print(keggplots[["KEGG_Dot"]][[cell_id]])
    dev.off()

    pdf(paste0("Plots/K7M2_2/",
               cell_id,
               "_KEGGenrichmentbar2.pdf"),
        height = 6,
        width = 12)
    print(keggplots[["KEGG_Bar2"]][[cell_id]])
    dev.off()

  } else {
    # notify if a cell type has less than 20 cells (excluded from analyses)
    message("Insufficent cell number for ", cell_id)
  }
  assign(cell_id, cell_type)
}
```

```{r}
sessionInfo()
```

