---
title: "A673 Tumor+Immune scRNAseq Analysis with HSV reference"
author: "Emily Franz"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 5
    number_sections: false
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  echo = TRUE,
  cache = TRUE,
  collapse = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  cache.lazy = FALSE)
```

```{r lib, cache = FALSE}
library(Seurat)
library(tidyverse)
library(ggsci)
library(rrrSingleCellUtils)
library(ggplot2)
library(patchwork)
source("scSeurat-addhsv.R")

# Set the random generator seed so that results are reproducible.
set.seed(153)

if(!file.exists("Data")){
  dir.create("Data")
}
```

## Tumor Samples - Human
Load all the different A673 datasets, note that HSV genes included but filter out mouse

```{r load, fig.height=5, fig.width=5}
sample_names <- c('S0207_HSV_exon_edits',
                  'S0208_HSV_exon_edits',
                  'S0209_HSV_exon_edits',
                  'S0210_HSV_exon_edits')

sample_labels <- c("1 - Control",
                   "2 - oHSV",
                   "3 - Trabectedin",
                   "4 - oHSV+Trabectedin")

# correlate sample names and sample labels
names(sample_labels) <- sample_names

sample_titles <- c("Control",
                   "oHSV",
                   "Trabectedin",
                   "Combination")
names(sample_titles) <- sample_names

# set max ncounts (will come back to adjust after vlnplot made, start at 1000000)
nCount_RNA_max <- list(1000000,
                       1000000,
                       1000000,
                       1000000)
names(nCount_RNA_max) <- sample_names

# set min for ncounts (will come back to adjust after vlnplot made, start at 0)
nCount_RNA_min <- c(0, 0, 0, 0)
names(nCount_RNA_min) <- sample_names

# set % mt genes (will come back to adjust after vlnplot made, start at 50)
mt_max <- c(50, 50, 50, 50)
names(mt_max) <- sample_names

for (item in sample_names) {
  obj <- tenx_load_qc(paste("/gpfs0/home2/gdrobertslab/lab/Counts/",
                            item,
                            "/filtered_feature_bc_matrix/",
                            sep = ""),
                      species_pattern = "^hg19_",
                      human_hsv = "TRUE")
  plot1 <- FeatureScatter(obj,
                          feature1 = "nCount_RNA",
                          feature2 = "percent.mt") +
    theme(legend.position="none")
  
  plot2 <- FeatureScatter(obj,
                          feature1 = "nCount_RNA",
                          feature2 = "nFeature_RNA")  +
    theme(legend.position="none")
  
  print(plot1 + plot2 +
          plot_annotation(title = sample_titles[[item]]))
  
  obj <- subset(obj,
                subset = 
                  nCount_RNA < nCount_RNA_max[[item]] &
                  nCount_RNA > nCount_RNA_min[[item]] &
                  percent.mt < mt_max[[item]])
  # make each sample an individual Seurat obj without overwriting
  assign(sample_titles[[item]], obj) 
}
```

Look at HSV transcripts vs mt transcripts and HSV transcripts vs nCounts.

oHSV-treated sample:

```{r scatter_hsv, fig.height=5, fig.width=10}
plot1_hsv <- FeatureScatter(oHSV,
                            feature1 = "nCount_RNA",
                            feature2 = "percent.hsv") +
  theme(legend.position="none")

plot2_hsv <- FeatureScatter(oHSV,
                            feature1 = "percent.mt",
                            feature2 = "percent.hsv") +
  theme(legend.position="none")

plot1_hsv +
  plot2_hsv +
  plot_annotation(title = 'oHSV-treated Tumor')
```

Combination-treated sample:

```{r scatter_combo, fig.height=5, fig.width=10}
plot1_hsv <- FeatureScatter(Combination,
                            feature1 = "nCount_RNA",
                            feature2 = "percent.hsv") +
  theme(legend.position="none")

plot2_hsv <- FeatureScatter(Combination,
                            feature1 = "percent.mt",
                            feature2 = "percent.hsv") +
  theme(legend.position="none")

plot1_hsv +
  plot2_hsv +
  plot_annotation(title = 'Combination-treated Tumor')
```

Using the previous violin plots, choose cutoffs to proceed with for processing.

```{r load_final}
# set max ncounts 
nCount_RNA_max <- list(60000,
                       45000,
                       60000,
                       20000)
names(nCount_RNA_max) <- sample_names

# set min for ncounts 
nCount_RNA_min <- c(800,
                    800,
                    800,
                    800)
names(nCount_RNA_min) <- sample_names

# set max ncounts 
nFeature_RNA_max <- list(8500,
                         7500,
                         7500,
                         8000)
names(nFeature_RNA_max) <- sample_names

# very high percent mt in tumor samples
mt_max <- c(40,
            40,
            40,
            40)
names(mt_max) <- sample_names

# Load Seurat objects and subset based on the above cutoffs
for (item in sample_names) {
  obj <- tenx_load_qc(paste("/gpfs0/home2/gdrobertslab/lab/Counts/",
                            item,
                            "/filtered_feature_bc_matrix/",
                            sep = ""),
                      species_pattern = "^hg19_|^mm10_",
                      mixed_hsv = "TRUE",
                      violin_plot = FALSE)
  
  obj <- subset(obj,
                subset = 
                  nCount_RNA < nCount_RNA_max[[item]] &
                  nCount_RNA > nCount_RNA_min[[item]] &
                  percent.mt < mt_max[[item]])

  obj$treatment <- sample_labels[[item]]

  # Make each sample an individual Seurat obj without overwriting
  assign(sample_titles[[item]], obj)
}
```

# Normalize and Process

```{r process, fig.height=3, fig.width=3, dependson="load_final"}
#### Merge ####
A673_hsv_rep2 <- merge(Control,
                  y = c(oHSV,
                        Trabectedin,
                        Combination),
                  add.cell.ids = c("Control",
                                   "oHSV",
                                   "Trabectedin",
                                   "Combination"),
                  project = "Trabectedin plus HSV1716 in A673 - repeat")

#### Normalize, scale, process, cluster ####
A673_hsv_rep2 <- NormalizeData(A673_hsv_rep2) %>%
  ScaleData() %>%
  FindVariableFeatures()

#### Variable Features ####
# Use Elbow Plots to determine the PCs to use
A673_hsv_rep2 <- RunPCA(A673_hsv_rep2,
                   features = VariableFeatures(object = A673_hsv_rep2))

ElbowPlot(A673_hsv_rep2,
          ndims = 30)

# Clean up environment
rm(Control, oHSV, Trabectedin, Combination)
```

### Clustering Resolution {.tabset}

#### Res0.1
```{r umap, fig.height=5, fig.width=10, dependson=c("load_final","process")}
A673_hsv_rep2 <- FindNeighbors(A673_hsv_rep2, dims = 1:9) %>%
  FindClusters(resolution = 0.1) %>%
  RunUMAP(dims = 1:9)

DimPlot(A673_hsv_rep2,
        reduction = "umap",
        split.by = "treatment",
        pt.size = 1,
        label = T) +
  coord_fixed()
```

#### Res0.2
```{r, umap2, fig.height=5, fig.width=10, dependson="umap"}
A673_hsv_rep2 <- FindClusters(A673_hsv_rep2,
                         resolution = 0.2) %>%
  RunUMAP(dims = 1:9)

DimPlot(A673_hsv_rep2,
        reduction = "umap",
        split.by = "treatment",
        pt.size = 1,
        label = T) +
  coord_fixed()
```

#### Res0.3
```{r, umap3, fig.height=5, fig.width=10, dependson="umap"}
A673_hsv_rep2 <- FindClusters(A673_hsv_rep2,
                         resolution = 0.3) %>%
  RunUMAP(dims = 1:9)

DimPlot(A673_hsv_rep2,
        reduction = "umap",
        split.by = "treatment",
        pt.size = 1,
        label = T) +
  coord_fixed()
```

# Downsample
Downsample so there is the same number of cells in each treatment group.

```{r downsample, fig.height=5, fig.width=10, dependson=c("load_final","process","umap")}
# Subset the highest number that makes all treatment groups equal in cell number
## Set seed to ensure reproducible results
set.seed(153)
Idents(A673_hsv_rep2) <- A673_hsv_rep2$treatment
A673_hsv_rep2_even <- subset(A673_hsv_rep2,
                        downsample = table(A673_hsv_rep2$treatment) %>%
                          min())

# Check that worked - all treatment groups should have same cell number
## Extract meta data
library(data.table)
A673_hsv_rep2_even.d <- A673_hsv_rep2_even@meta.data %>%
  as.data.table

## Resulting table has one "row" per cell - count
A673_hsv_rep2_even.d[, .N, by = "treatment"]

# Plot
DimPlot(A673_hsv_rep2_even,
        split.by = "treatment") +
  coord_fixed()
```

```{r save, dependson= c("load_final","process","umap","downsample","singlr")}
save(A673_hsv_rep2_even, file  = "Data/tumor-immune-A673_hsv_rep2_even-labeled.RData")
```

```{r}
sessionInfo()
```
